### DISCUSSION TITLE
New `r` schema for requests in Data Store

### DISCUSSION THREAD

**Discussion by alexiglesias93** [finsweet/wized-roadmap#1] [RFCs]
## The problem
Currently, the requests schema in the data store looks something like this:
```typescript
const dataStore = {
  r: {
    1: {
      d: {},
      $: {
        isRequesting: false,
        hasRequested: true,
        status: 200,
        statusText: 'OK',
      },
    },
  },
};
```

I think this presents a couple UX issues when accessing the data:

### Lack of autocomplete in the new Data Field editor
Typing `r.1` is not valid JavaScript. Users have to use brackets notation like `r[1]`. On top of that, IntelliSense in the code editor doesn't display the available options.
For instance, when typing `v.` you get all the existing variables as suggestions in the editor:
![image](https://github.com/finsweet/wized-rfcs/assets/61051948/25d3bb80-d474-4dab-ba67-5d7deb793421)

But when typing `r.` you don't get any suggestions because the JavaScript language expects bracket notations:
![image](https://github.com/finsweet/wized-rfcs/assets/61051948/d53e94cf-47c2-41b3-ace1-cd1e2555829a)

It's not until the user types `r[1].` that he gets the correct autocomplete suggestions:
![image](https://github.com/finsweet/wized-rfcs/assets/61051948/61bc9214-3167-44c7-8f75-a6050e832b6e)

### Confusing `$` and `d` properties
I think it's not very straightforward for users to understand what `$` and `d` stand for.
`$` stands for state, and `d` stands for data.
But there aren't any real reasons why it's needed to be defined this way, as the request state and the data object will never overlap.

### Improving autocomplete
Autocomplete for requests can be improved a lot by just doing a simple change: using the request name as the accessor instead of the request number.
```typescript
const dataStore = {
  r: {
    request_name: {
      // ...
    },
  },
};
```

It would allow users to see the autocomplete suggestions:
![image](https://github.com/finsweet/wized-rfcs/assets/61051948/ff38787a-a8bd-496c-a854-5a3be7c8a814)

And on top of that, it would provide better context when developing. If you read some code, it's easier to understand what's going on if you read `r.login_user` rather than `r[2]`. The request number doesn't give any context at all.

The benefits are clear, but I believe there's one downside: we'll have to force request names to be JS friendly.
Same as we do for variable names, we'll have to add some rules that automatically rewrite a request name, like removing spaces, dashes and other invalid characters:
- `Login User` -> `Login_User`
- `fetch-The Data` -> `fetchThe_Data`
- `i'm loading` -> `im_Loading`
- `loginUser` -> `loginUser` (no renaming needed if there are no invalid characters)
- etc...

I don't think it's a big deal, in fact it might even help people keep a consistent naming and organization in their project. But maybe not everyone agrees.

### Reworking the request schema
As I mentioned, I think that `$` and `d` don't provide much context to users.
There's no real reason why we need to keep both namespaces separate and with those short names. I propose we update the schema to be:
```typescript
const dataStore = {
  r: {
    request_name: {
      data: {},
      isRequesting: false,
      hasRequested: false,
      status: 200,
      statusText: 'OK',
      ok: true,
    },
  },
};
```
With this change, we unify all the possible request data under a single namespace, giving more context to the user at a first glance:
![image](https://github.com/finsweet/wized-rfcs/assets/61051948/9423695b-0218-4a70-9633-c8310876f18b)

In addition, I propose including the standardized [Response.ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) property from browsers, as it's a very handy shorthand to check if the response was successful:
```typescript
// This:
r.login_user.status === 200 || r.login_user.status === 201 || r.login_user.status === 202

// Can be replaced with this:
r.login_user.ok
```

**Comment by theflowagency** [comment]
This makes sense to me at least Alex. If anything, it makes the code more readable which therefore makes it easier to understand for the less technical user. Since Wized will be attracting more and more people who are not used to seeing or using code, anything to simplify the construction of code blocks in the new editor will be advantageous.

**Comment by dtavdi** [comment]
super cool!

**Comment by benoiteveillard** [comment]
100% agreed! 
According to the "downside", as you said _it might even help people keep a consistent naming and organization in their project_ so I do think it is **not a downside**.

As you said (or Joe?) earlier this year, Webflow helped a lot of people to understand HTML & CSS (including me) and you want Wized help people to better understand JS. Better naming convention (as a traditional developer would do) is part of it. So I think you will meet your objective if people adopt these kinds of reflexes thanks to Wized automatic standardization.

And the standardized Response.ok property seems also a great idea

**Comment by alexiglesias93** [comment]
Closing this as it was already shipped.
